BrainPortal/lib/s3_connection.rb

```diff
@@ -19,52 +18,244 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
-
-require 'aws/s3'
+require 'aws-sdk-s3'
+require 'fileutils'
 
 # A handler connections to Amazon's S3 service.
 class S3Connection
 
   Revision_info=CbrainFileRevision[__FILE__] #:nodoc:
 
-  attr_accessor :base,:service, :bucket, :s3object
+  attr_accessor :resource, :bucket_name, :path_name, :region
 
   # Establish a connection handler to S3
-  def initialize(access_key, secret_key)
-    @base     = Base
-    @service  = Service
-    @bucket   = Bucket
-    @s3object = S3Object
-
-    @base.establish_connection!(
-      :access_key_id     => (access_key || "blah"),
-      :secret_access_key => (secret_key || "blah")
-    )
+  def initialize(access_key, secret_key, bucket_name, path_start, 
+                 region='us-east-2',
+                 endpoint="http://s3.us-east-2.amazonaws.com")
+    credentials = Aws::Credentials.new(access_key,secret_key)
+    @bucket_name = bucket_name
+    @region = region
+    @client = Aws::S3::Client.new(credentials: credentials,
+                                  region: @region,
+                                  endpoint: endpoint)
+    @resource = Aws::S3::Resource.new(client: @client)
+    @bucket = @resource.bucket(@bucket_name)
+    @path_name = path_start
+    @endpoint = endpoint
   end
-
+  
+  # Method to translate Mime types to File types
+  # Redo more generally
+  def translate_content_type_to_ftype(ct)
+    ct == "application/x-directory" ? :directory : :regular
+  end
+      
+  # Method to ensure that the paths sent start in the right place
+  def clean_starting_folder_path(path)
+    path_end = path
+    
+    if not path_end.to_s.starts_with? @path_name
+      return Pathname.new(File.join(@path_name, path_end))
+    else
+      return Pathname.new(path_end)
+    end
+  end
+  
   def execute_on_s3 #:nodoc:
      yield self
   end
 
-  # Create a bucket on the current connection.
-  def create_bucket(bucket_name)
+  # Create a bucket on the current connection. Not used, included for completeness
+  def self.create_bucket(bucket_name)
     @bucket.create(bucket_name)
   end
 
   # Returns true if the connection is alive
+  # Need to figure out how to return the actual error message so that one can debug
   def connected?
-    @base.connected?
+    begin
+      testVar = @resource.client.list_objects(bucket: @bucket_name, delimiter: "/")
```

Comment:
Isn't it costly to list a whole bunch of objects (can be thousands!) just to test the alivemness?

Reply:
Can only be 1000 and this is only one API call to the Rest interface, it was the preferred way that was described by the developers of aws-s3-sdk to do this.  You actually have to make an api call for the gem to connect to the s3 store.  So this should not present significant overhead.